File: visualization/backend/astrolabe/models/node.py
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional


class ProofStatus(Enum):
    PROVEN = "proven"
    SORRY = "sorry"
    ERROR = "error"
    UNKNOWN = "unknown"


@dataclass
class NodeMeta:
    """User-editable properties via UI, stored in .astrolabe/meta.json"""

    # Display
    label: Optional[str] = None
    size: Optional[float] = None
    shape: Optional[str] = None
    effect: Optional[str] = None

    # Position related (pinned still kept, as it's a node property, not position data)
    # Note: Actual position data is stored in canvas.json, not in meta.json
    pinned: bool = False

    # Content
    notes: Optional[str] = None
    tags: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        """Output only non-empty fields"""
        result = {}
        if self.label:
            result["label"] = self.label
        if self.size is not None:
            result["size"] = self.size
        if self.shape:
            result["shape"] = self.shape
        if self.effect:
            result["effect"] = self.effect
        if self.pinned:
            result["pinned"] = self.pinned
        if self.notes:
            result["notes"] = self.notes
        if self.tags:
            result["tags"] = self.tags
        return result

    @classmethod
    def from_dict(cls, data: dict) -> "NodeMeta":
        """Create NodeMeta from dict

        Note: For backward compatibility, ignores old fields like color, texStatement, texProof
        """
        return cls(
            label=data.get("label"),
            size=data.get("size"),
            shape=data.get("shape"),
            effect=data.get("effect"),
            pinned=data.get("pinned", False),
            notes=data.get("notes"),
            tags=data.get("tags", []),
        )


@dataclass
class Node:
    """
    Astrolabe Node

    Data sources:
    - Lean files (via .ilean parsing): id, name, kind, file_path, line_number, status, references
    - .astrolabe/meta.json (user edited): meta

    Note: Source code content is fetched on-demand via readFile API, not stored in node
    """

    # === From Lean (source of truth, read-only) ===
    id: str
    name: str
    kind: str  # theorem, def, lemma, structure...
    file_path: str
    line_number: int
    status: ProofStatus = ProofStatus.UNKNOWN
    references: list[str] = field(default_factory=list)

    # === Statistics fields (computed) ===
    depends_on_count: int = 0  # How many nodes this node depends on
    used_by_count: int = 0     # How many nodes reference this node
    depth: int = 0             # Dependency chain depth (0=leaf node)

    # === Default styles (set based on kind) ===
    default_color: str = "#888888"
    default_size: float = 1.0
    default_shape: str = "sphere"

    # === Internal use (not serialized) ===
    _full_content: str = ""  # Full content, for dependency analysis and sorry detection

    # === From .astrolabe/meta.json (user editable) ===
    meta: NodeMeta = field(default_factory=NodeMeta)

    def to_dict(self) -> dict:
        """Serialize for frontend"""
        result = {
            "id": self.id,
            "name": self.name,
            "kind": self.kind,
            "filePath": self.file_path,
            "lineNumber": self.line_number,
            "status": self.status.value,
            "references": self.references,
            "dependsOnCount": self.depends_on_count,
            "usedByCount": self.used_by_count,
            "depth": self.depth,
            "defaultColor": self.default_color,
            "defaultSize": self.default_size,
            "defaultShape": self.default_shape,
            "meta": self.meta.to_dict(),
        }
        return result

File: visualization/src/types/node.ts
/**
 * Astrolabe Graph Types
 * Unified Node and Edge type definitions
 *
 * Data sources:
 * - Lean files (source of truth): id, name, kind, filePath, lineNumber, content, status, references
 * - .astrolabe/meta.json (user edited): meta
 */

// ============================================
// 基础枚举类型
// ============================================

export type NodeKind =
  | "theorem"
  | "lemma"
  | "definition"
  | "proposition"
  | "corollary"
  | "axiom"
  | "conjecture"
  | "example"
  | "structure"
  | "class"
  | "instance"
  | "inductive"
  | "custom";  // Virtual node

export type ProofStatus =
  | "proven"  // Found in Lean, no sorry
  | "sorry"   // Found in Lean, has sorry
  | "error"   // Has errors
  | "stated"  // Types that don't require proof (definition, axiom, etc.)
  | "loading" // Awaiting diagnostic verification (cached status)
  | "unknown"; // Referenced but not defined

// ============================================
// NodeMeta type (from .astrolabe/meta.json)
// ============================================

export interface NodeMeta {
  // Display
  label?: string;
  size?: number;
  shape?: string;
  effect?: string;

  // Position
  position?: [number, number, number];
  pinned?: boolean;

  // Content
  notes?: string;
  tags?: string[];
}

// ============================================
// Node type (matches Python Node.to_dict() output)
// ============================================

export interface Node {
  // === From Lean (source of truth, read-only) ===
  id: string;
  name: string;
  kind: NodeKind;
  filePath: string;
  lineNumber: number;
  status: ProofStatus;
  references: string[];

  // === Statistics fields (calculated) ===
  dependsOnCount: number;
  usedByCount: number;
  depth: number;

  // === Default styles (obtained from theme based on kind) ===
  defaultColor: string;
  defaultSize: number;
  defaultShape: string;

  // === From .astrolabe/meta.json (user editable) ===
  meta: NodeMeta;
}

// ============================================
// EdgeMeta type (from .astrolabe/meta.json)
// ============================================

export interface EdgeMeta {
  style?: string;  // 'solid' | 'dashed' | 'polyline'
  effect?: string;
  notes?: string;
}

// ============================================
// Edge type (matches Python Edge.to_dict() output)
// ============================================

export interface Edge {
  id: string;
  source: string;
  target: string;
  fromLean: boolean;
  visible?: boolean;

  // === Default styles (obtained from theme based on fromLean) ===
  defaultColor: string;
  defaultWidth: number;
  defaultStyle: string;  // 'solid' | 'dashed' | 'polyline'

  meta: EdgeMeta;
}

// ============================================
// API 请求/响应类型
// ============================================

export interface NodeMetaUpdate {
  label?: string;
  size?: number;
  shape?: string;
  effect?: string;
  position?: [number, number, number];
  pinned?: boolean;
  notes?: string;
  tags?: string[];
}

export interface ProjectStats {
  total_nodes: number;
  total_edges: number;
  by_kind: Record<string, number>;
  by_status: Record<string, number>;
}

export interface ProjectData {
  path: string;
  nodes: Node[];
  edges: Edge[];
  stats?: ProjectStats;
}

export interface ProjectStatus {
  exists: boolean;
  hasLakefile: boolean;
  hasLakeCache: boolean;
  usesMathlib: boolean;
  leanFileCount: number;
  needsInit: boolean;
  notSupported: boolean;
  message: string;
}
